rust/lance/src/index/vector/ivf/shuffler.rs

```diff
@@ -0,0 +1,177 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Disk-based shuffle a stream of [RecordBatch] into each IVF partition.
+//! 
+//! 1. write the entire stream to a file
+//! 2. count the number of rows in each partition
+//! 3. with memory buffering, read the file and write to each partition
+
+use arrow_array::UInt32Array;
+use arrow_array::cast::{as_primitive_array, as_fixed_size_list_array};
+use arrow_array::types::{UInt32Type, UInt64Type, UInt8Type};
+use arrow_schema::{Schema as ArrowSchema, Field as ArrowField, DataType};
+use futures::{StreamExt, stream};
+use kdam::BarExt;
+use lance_core::datatypes::Schema;
+use lance_core::io::reader::batches_stream;
+use lance_core::io::{RecordBatchStream, FileWriter, FileReader, ReadBatchParams};
+
+
+use lance_core::io::object_store::ObjectStore;
+use lance_core::{Error, Result, ROW_ID};
+use lance_index::vector::{PART_ID_COLUMN, PQ_CODE_COLUMN};
+use object_store::path::Path;
+use snafu::{location, Location};
+use tempfile::TempDir;
+
+const UNSORTED_BUFFER: &str = "unsorted.lance";
+
+fn get_temp_dir() -> Result<Path> {
+    let dir = TempDir::new()?;
+    let tmp_dir_path = Path::from_filesystem_path(dir.path()).map_err(|e| Error::IO {
+        message: format!("failed to get buffer path in shuffler: {}", e),
+        location: location!(),
+    })?;
+    Ok(tmp_dir_path)
+}
+
+pub struct IvfShuffler {
+    pub num_partitions: u32,
+
+    pub output_dir: Path,
+
+    pub schema: Schema,
+}
+
+impl IvfShuffler {
+    pub async fn dump_unsorted_stream(
+        &self,
+        data: impl RecordBatchStream + Unpin + 'static,
+    ) -> Result<()> {
+        let object_store = ObjectStore::local();
+        let path = self.output_dir.child(UNSORTED_BUFFER);
+        let writer = object_store.create(&path).await?;
+
+        let mut file_writer = FileWriter::with_object_writer(
+            writer,
+            self.schema.clone(),
+            &Default::default(),
+        )?;
+
+        let mut data = Box::pin(data);
+
+        while let Some(batch) = data.next().await {
+            file_writer.write(&[batch?]).await?;
+        }
+
+        file_writer.finish().await?;
+
+        Ok(())
+    }
+
+    pub async fn count_partition_size(&self) -> Result<Vec<u64>> {
+        let object_store = ObjectStore::local();
+        let path = self.output_dir.child(UNSORTED_BUFFER);
+        let reader = FileReader::try_new(&object_store, &path).await?;
+
+        let mut partition_sizes = vec![0; self.num_partitions as usize];
+
+        // stream out just the partition id column
+        let schema = ArrowSchema::new(vec![
+            ArrowField::new(PART_ID_COLUMN, DataType::UInt32, false)
+        ]);
+
+        let lance_schema = Schema::try_from(&schema)?;
```

Comment:
This creating the lance schema like this isn't safe. You should instead take the schema off of the reader and then project the `PART_ID_COLUMN`.

The schema passed into `read_batch()` needs to have field ids that match the original schema. You are getting incorrect data because it is just pulling the first field in the schema (`field_id = 0`, which is `_rowid`) and reinterpretting that as u32.

The reason this was sensitive to the number of partitions is you were doing `(batch_i << 32)` to generate the row id, so those first 32 bits were turning into the first value of the misread part ids.